File: src/main/java/net/eduard/api/lib/manager/TimeManager.java
package net.eduard.api.lib.manager;

import org.bukkit.Bukkit;
import org.bukkit.plugin.Plugin;
import org.bukkit.scheduler.BukkitTask;

import net.eduard.api.lib.modules.BukkitTimeHandler;


/**
 * Gerenciador de Timers e Tasks
 *
 * @author Eduard
 */

public class TimeManager extends EventsManager implements Runnable, BukkitTimeHandler {
     transient long taskDuration = 20;
     transient long taskStart;
     transient BukkitTask taskUsed;
    public boolean existsTask() {
        return taskUsed != null;
    }
    /**
     * Desliga o Timer/Delay criado
     */
    public void stopTask() {
        if (existsTask()) {
            taskUsed.cancel();
            Bukkit.getScheduler().cancelTask(taskUsed.getTaskId());
            taskUsed = null;
        }
    }
    public void run() {
    }
    public Plugin getPluginConnected() {
        return getPlugin();
    }
    public boolean isRunning() {
        return existsTask() && Bukkit.getScheduler().isCurrentlyRunning(taskUsed.getTaskId());
    }

    public TimeManager() {
        this(1);
    }
    public TimeManager(long ticks) {
        this.taskDuration = ticks;
    }
    public TimeManager(int seconds) {
        this.taskDuration = 20L * seconds;
    }

    /**
     * Cria um Sync Delay
     *
     * @return BukkitTask
     */
    public BukkitTask syncDelay() {
        taskUsed = newTask(taskDuration, false, false, this);
        taskStart = System.currentTimeMillis();
        return taskUsed;
    }

    /**
     * Cria um Sync Timer
     *
     * @return Timer
     */
    public BukkitTask syncTimer() {
        taskUsed = newTask(taskDuration, true, false, this);
        taskStart = System.currentTimeMillis();
        return taskUsed;
    }

    /**
     * Cria um Async Timer
     *
     * @return Timer
     */
    public BukkitTask asyncTimer() {
        taskUsed = newTask(taskDuration, true, true, this);
        taskStart = System.currentTimeMillis();
        return taskUsed;
    }

    /**
     * Cria um Async Delay
     *
     * @return Delay
     */
    public BukkitTask asyncDelay() {
        taskUsed = newTask(taskDuration, false, true, this);
        taskStart = System.currentTimeMillis();
        return taskUsed;
    }
}

File: src/main/java/net/eduard/api/lib/manager/TimeManager.kt
package net.eduard.api.lib.manager

import org.bukkit.Bukkit
import org.bukkit.plugin.Plugin
import org.bukkit.scheduler.BukkitTask

import net.eduard.api.lib.modules.BukkitTimeHandler
import net.eduard.api.lib.modules.Mine

/**
 * Classe que Herda EventsManager, como se fosse um BukkitRunnable que
 * pode aceitar Eventos Também, utiliza a classe BukkitTimeHandler para
 * interagir com o BukkitScheduler escrevendo apenas Aliases
 *
 * @author Eduard
 */
open class TimeManager(var time: Long) : EventsManager(), Runnable,
    BukkitTimeHandler {

    constructor() : this(20)

    /**
     * Tempo do Inicio do Delay/Timer
     */

    var startedTime: Long = 0


    /**
     * BukkitTask criado apartir das aliases internas
     */
    @Transient
    lateinit var task: BukkitTask

    /**
     * Verifica se este Timer/Delay esta em execução
     */
    val isRunning: Boolean
        get() = existsTask() &&
                Bukkit.getScheduler().isCurrentlyRunning(task.taskId)


    /**
     * Se for Timer executa o run() repetidamente a cada X tempo,
     * se for delay executa o run() apenas uma vez depois de X Tempo
     */
    override fun run() {}

    /**
     * Retorna um Plugin que Interagiu com esta classe primeiro,
     * geralmente retorna EduardAPIMain
     */
    override fun getPluginConnected(): Plugin {
        return plugin
    }

    /**
     * Cria um Sync Delay
     *
     *
     * @return Delay
     */
    fun syncDelay(): BukkitTask {
        task = newTask(time, false, false, this)
        startedTime = System.currentTimeMillis()
        return task
    }

    /**
     * Cria um Sync Timer
     *
     *
     * @return Timer
     */
    fun syncTimer(): BukkitTask {
        task = newTask(time, true, false, this)
        startedTime = System.currentTimeMillis()
        return task
    }

    /**
     * Cria um Async Timer
     *
     *
     * @return Timer
     */
    fun asyncTimer(): BukkitTask {
        task = newTask(time, true, true, this)
        startedTime = System.currentTimeMillis()
        return task
    }

    /**
     * Cria um Async Delay
     *
     *
     * @return Delay
     */
    fun asyncDelay(): BukkitTask {
        task = newTask(time, false, true, this)
        startedTime = System.currentTimeMillis()
        return task
    }

    /**
     *
     * @return Se ligou um Timer/Delay, porem pode já estar finalizado
     */
    fun existsTask(): Boolean {
        return this::task.isInitialized
    }

    /**
     * Desliga o Timer/Delay criado
     */
    fun stopTask() {
        if (existsTask())
            task.cancel()

    }


}
